packages/vite/src/node/plugins/asset.ts

```diff
@@ -241,6 +243,39 @@ async function fileToBuiltUrl(
   return url
 }
 
+function specialHexEncode(match: string) {
+  // Browsers tolerate these characters, and they're frequent
+  switch (match) {
+    case '%20':
+      return ' '
+    case '%3D':
+      return '='
+    case '%3A':
+      return ':'
+    case '%2F':
+      return '/'
+    default:
+      return match.toLowerCase() // compresses better
+  }
+}
+
+const doubleQuoteRE = /"/g
+const whitespaceRE = /\s+/g
+const urlHexPairsRE = /%[\dA-F]{2}/g
+
+// Adopted from https://github.com/tigt/mini-svg-data-uri
+function uriEncodeSvg(content: string) {
+  const normalizedContent = content
+    .trim()
+    .replace(whitespaceRE, ' ')
+    .replace(doubleQuoteRE, "'")
+
+  return encodeURIComponent(normalizedContent).replace(
+    urlHexPairsRE,
+    specialHexEncode
+  )
+}
```

Comment:
I did not go with `mini-svg-data-uri` precisely because it does some micro-optimizations that aren't strictly necessary namely conversion from `%20` to ` `, which `toSrcset` has to undo as its _only_ function. 

However, since the team's sentiment seems to be towards using the upstream package I switched to that. I had to always use `toSrcset` though, because there no way to tell whether the Data URI string is going to be used in `srcSet` when importing into JavaScript.

Reply:
> I had to always use `toSrcset` though, because there no way to tell whether the Data URI string is going to be used in `srcSet` when importing into JavaScript.

That's fine :+1: 

If the detection via context of usage isn't available better that it works with all cases where white-space would otherwise cause a breakage (there is more than just `srcSet` where SVG can be provided and this would break without encoding spaces).

For anyone that outputs static HTML/CSS, they can probably post-process if they want to convert `%20` to ` ` (space) when valid. It can make for decent savings on larger SVG uncompressed(it's often larger than base64 with encoded spaces), but most of the time compression will be in use anyway which closes the gap notably.

For example, an SVG of 1,164 bytes url-encoded (but spaces retained) is 590 bytes gzip level 6, and the space encoded version is 1,494 bytes but only 623 bytes via gzip 6. Base64 equivalent 1,474 bytes uncompressed and 863 bytes gzip 6. Original SVGO optimized SVG is 1,084 bytes as a file (561 bytes gzip 6).